Implement LLM response streaming

dlqqq opened the pull request:
## Description

This is a large PR that implements LLM response streaming in Jupyter AI. LangChain LLM classes that implement the `_stream()` or `_astream()` methods can render the response incrementally in chunks, which allows users to view the response being built token-by-token.

Also fixes #858 by adding a version ceiling on `faiss-cpu`.

### Demo of single-user scenario

https://github.com/jupyterlab/jupyter-ai/assets/44106031/c54f3294-2a99-46af-b336-c89d99131ad8

### Demo of multi-user scenario, showing stream completeness & consistency

https://github.com/jupyterlab/jupyter-ai/assets/44106031/972754d1-8123-45b6-892c-2802160d41f5

## Extended description (for developers & reviewers)

- The `"Generating response..."` pending message is still shown to the user while waiting to receive the first chunk from the LLM.
- How chat history is retrieved & managed by the client has been fully re-worked; the initial `ConnectionMessage` object streamed to the client by the server extension on connection now includes a `ChatHistory` object, which contains the `ChatHistory`.
  - Clients no longer need to call the `GetHistory` REST API endpoint to retrieve the history separately; the entire chat history can now be obtained by listening to the Chat websocket connection.
  - This change was made to ensure that new clients connecting while Jupyter AI is streaming a response do not "miss" any chunks that are being actively streamed while joining. **Since the chat history is received as soon as the WebSocket connection is established, clients should not miss any chunks even if they join mid-stream.** This is demonstrated in the "Demo of multi-user scenario" section above.
- This PR also introduces a `jupyter_ai_test` package that includes a `TestProvider` and `TestStreamingProvider`.
  - **For reviewers:** `jupyter_ai_test` can be installed in your dev environment simply by running `jlpm dev-uninstall && jlpm dev-install`.
  - Since this package is not listed under `.jupyter_releaser.toml`, this package will not be released to NPM or PyPI. It is intended for testing in local development workflows only.

## Follow-up work

- The "Include selection" and "Replace selection" checkboxes need to be removed **prior to the next release,** as they likely do not work anymore. To avoid a breaking change, they must be replaced by alternative features in the UI when removed:
  - "Include selection": This should be implemented by replacing the current "Send" button with a dropdown button that allows users to send a prompt with a text/cell selection.
  - "Replace selection": This should be implemented by:
    1. Supporting replacing text selection in the code action toolbar
    2. Adding a hamburger menu on each message that contains an option for replacing the current text/cell selection with the entire Markdown contents of a message.
  - I argue **this should be done in a future PR prior to the next release** however, as these features lie out of the scope of this PR.

3coins commented on packages/jupyter-ai/src/chat_handler.ts:
```diff
@@ -106,23 +96,84 @@ export class ChatHandler implements IDisposable {
+        if (newMessage.stream_complete) {
+          streamMessage.complete = true;
+        }
+        this._messages = [...this._messages];
```
Why is this required?

dlqqq replied:
This is done so consumer React components re-render when directly calling `setMessages()` on the returned `history.messages` object that is emitted in the `historyChanged` signal. I added a comment at the top of the `switch` block for context:


```ts
    // append message to chat history. this block should always set `_messages`
    // or `_pendingMessages` to a new array instance rather than modifying
    // in-place so consumer React components re-render.
    switch (newMessage.type) {
       ...
    }
```

3coins reviewed (commented):
@dlqqq 
Kudos on adding streaming so quickly, the UX is much better with these changes. Added some minor suggestions, but looks good otherwise. 
Noticed that /ask is still using the non-streaming messages, do you plan to handle streaming for /ask in a separate PR?

3coins commented on packages/jupyter-ai-test/jupyter_ai_test/_version.py:
```diff
@@ -0,0 +1,4 @@
+# This file is auto-generated by Hatchling. As such, do not:
+#   - modify
+#   - track in version control e.g. be sure to add to .gitignore
+__version__ = VERSION = "0.1.0"
```
```suggestion
__version__ = VERSION = '0.1.0'
```

krassowski replied:
I think this file should just be deleted. It should not be committed in the first place.

dlqqq replied:
Thank you both for catching this. I've deleted this file in the latest revision.

JasonWeill commented on packages/jupyter-ai/jupyter_ai/history.py:
```diff
@@ -0,0 +1,40 @@
+        if self.size > self.k * 2:
+            self.messages.pop(0)
+
+    async def aadd_messages(self, messages: Sequence[BaseMessage]) -> None:
```
Is it common Python practice to name a method `foo` if it's synchronous and `afoo` if it's async? `aadd_messages` read like a typo at first, but if this naming convention is typical, then let's leave it as it is.

dlqqq replied:
I don't think there is a generally-accepted naming convention to distinguish sync methods from async methods. However, we have to define the method as `aadd_messages()` because that is expected to be available when used as a history object by LangChain. This can be verified from the `BaseChatMessageHistory` docs: https://api.python.langchain.com/en/latest/chat_history/langchain_core.chat_history.BaseChatMessageHistory.html#

JasonWeill reviewed (commented):
This looks and works great! Little to add beyond @3coins 's comments earlier. Also, could you please file issues to cover the proposed additional enhancements out of scope for this PR?

3coins commented on packages/jupyter-ai/jupyter_ai/chat_handlers/default.py:
```diff
@@ -1,10 +1,18 @@
+)
 from jupyter_ai_magics.providers import BaseProvider
-from langchain.chains import ConversationChain, LLMChain
 from langchain.memory import ConversationBufferWindowMemory
```
Is this still required?

dlqqq replied:
Great catch, thank you! I've removed all references of `ConversationBufferWindowMemory` in the latest revision.

dlqqq commented on packages/jupyter-ai-test/package.json:
```diff
@@ -0,0 +1,25 @@
+{
+  "name": "@jupyter-ai/test",
+  "version": "2.18.1",
```
The version of this test package should always match the version of `jupyter_ai`, since our version bumping script bumps all workspace packages (under the `packages/` directory) in lock-step. So it is fine to leave the version declared as `2.18.1` here.


Also, `_version.py` has been deleted as it should not be tracked in the source repo, per a separate review thread.

dlqqq commented on packages/jupyter-ai-test/README.md:
```diff
@@ -0,0 +1,58 @@
+# jupyter_ai_test
+
+`jupyter_ai_test` is a Jupyter AI module, a package
+that registers additional model providers and slash commands for the Jupyter AI
+extension.
```
I've updated the description to provide more specific context:

```
`jupyter_ai_test` is a Jupyter AI module that registers additional model
providers and slash commands for testing Jupyter AI in a local development
environment. This package should never published on NPM or PyPI.
```

dlqqq commented:
@3coins @JasonWeill Thank you both for the thoughtful review. I've addressed your comments in the latest revision.

> Noticed that /ask is still using the non-streaming messages, do you plan to handle streaming for /ask in a separate PR?

Yes. There were some issues on the `langchain-ai/langchain` repo about LCEL streaming when using retrievers, so I decided it would be best to implement streaming for `/ask` in a separate, future PR.

3coins commented:
@dlqqq 
Here is a reference to building RAG with LCEL, but ok to do this in a separate PR.
https://python.langchain.com/v0.1/docs/use_cases/question_answering/chat_history/#tying-it-together

dlqqq commented:
Opened a new issue to track `/ask` streaming: https://github.com/jupyterlab/jupyter-ai/issues/863

wenmin-wu commented:
how to interrupt the generation?
